packages/blocks/src/api/registration.js

```diff
@@ -900,13 +924,199 @@ export const registerBlockBindingsSource = ( source ) => {
 		return;
 	}
 
+	if ( getFieldsList ) {
+		deprecated( 'getFieldsList in block bindings source', {
+			since: '6.9',
+			alternative: 'editorUI',
+			hint: 'The editorUI property provides more flexibility and better UI integration.',
+		} );
+	}
 	// Check the `getFieldsList` property is correct.
 	if ( getFieldsList && typeof getFieldsList !== 'function' ) {
 		// eslint-disable-next-line no-console
 		warning( 'Block bindings source getFieldsList must be a function.' );
 		return;
 	}
 
+	if ( editorUI ) {
+		if ( typeof editorUI === 'function' ) {
+			// If it's a function, we can't validate the returned object at registration time
+			// since it may depend on runtime context, so we just validate it's a function
+			return unlock( dispatch( blocksStore ) ).addBlockBindingsSource(
+				source
+			);
+		}
+		if ( typeof editorUI !== 'object' || Array.isArray( editorUI ) ) {
+			warning( 'EditorUI must be an object or a function' );
+			return;
+		}
+		const { mode, data } = editorUI;
+
+		if ( mode && ! [ 'dropdown', 'modal' ].includes( mode ) ) {
+			warning( 'EditorUI mode must be either "dropdown" or "modal"' );
+			return;
+		}
+
+		if ( mode === 'dropdown' ) {
+			if ( ! data || ! Array.isArray( data ) ) {
```

Comment:
Isn't it too limiting that `data` needs to be fully defined at the time of source registration? What if the list of fields dynamically changes upon a user interaction? For example, when user creates a new post meta, how it could get added to the dropdown?

Reply:
I'm trying to understand better the rationale between defining `editorUI` as a function with no validation, and defining as an object with a very strict validation at registration time. I feel like I'm missing a proper context here.

Reply:
> Isn't it too limiting that data needs to be fully defined at the time of source registration? What if the list of fields dynamically changes upon a user interaction? For example, when user creates a new post meta, how it could get added to the dropdown?

That is indeed a really good "what if". The idea in a future is to be able to create new post meta with the UI, and should be added to the dropdown dynamically. Let's change it.